crates/debug/src/transform/expression.rs

```diff
@@ -218,27 +218,65 @@ impl CompiledExpression {
-        addr_tr: &AddressTransform,
-        frame_info: Option<&FunctionFrameInfo>,
-        isa: &dyn TargetIsa,
-    ) -> Result<Vec<(write::Address, u64, write::Expression)>> {
+    pub fn build_with_locals<'a>(
+        &'a self,
+        scope: &'a [(u64, u64)], // wasm ranges
+        addr_tr: &'a AddressTransform,
+        frame_info: Option<&'a FunctionFrameInfo>,
+        isa: &'a dyn TargetIsa,
+    ) -> impl Iterator<Item = Result<(write::Address, u64, write::Expression)>> + 'a {
+        enum BuildWithLocalsResult<'a> {
+            Empty,
+            Simple(
+                Box<dyn Iterator<Item = (write::Address, u64)> + 'a>,
+                Vec<u8>,
+            ),
+            Ranges(
+                Box<dyn Iterator<Item = Result<(DefinedFuncIndex, usize, usize, Vec<u8>)>> + 'a>,
+            ),
+        }
+        impl Iterator for BuildWithLocalsResult<'_> {
+            type Item = Result<(write::Address, u64, write::Expression)>;
+            fn next(&mut self) -> Option<Self::Item> {
+                match self {
+                    BuildWithLocalsResult::Empty => None,
+                    BuildWithLocalsResult::Simple(it, code) => it
+                        .next()
+                        .map(|(addr, len)| Ok((addr, len, write::Expression(code.to_vec())))),
+                    BuildWithLocalsResult::Ranges(it) => it.next().map(|r| {
+                        r.map(|(func_index, start, end, code_buf)| {
+                            (
+                                write::Address::Symbol {
+                                    symbol: func_index.index(),
+                                    addend: start as i64,
+                                },
+                                (end - start) as u64,
+                                write::Expression(code_buf),
+                            )
+                        })
+                    }),
+                }
+            }
+        }
+
         if scope.is_empty() {
-            return Ok(vec![]);
+            return BuildWithLocalsResult::Empty;
         }
 
         // If it a simple DWARF code, no need in locals processing. Just translate
         // the scope ranges.
         if let [CompiledExpressionPart::Code(code)] = self.parts.as_slice() {
-            let result_scope = scope
-                .iter()
-                .map(|(wasm_start, wasm_end)| addr_tr.translate_ranges(*wasm_start, *wasm_end))
-                .flatten()
-                .map(|(addr, len)| (addr, len, write::Expression(code.to_vec())))
-                .collect();
-            return Ok(result_scope);
+            return BuildWithLocalsResult::Simple(
+                Box::new(
+                    scope
+                        .iter()
+                        .map(move |(wasm_start, wasm_end)| {
+                            addr_tr.translate_ranges(*wasm_start, *wasm_end)
+                        })
+                        .flatten(),
```

Comment:
```suggestion
                        .flat_map(move |(wasm_start, wasm_end)| {
                            addr_tr.translate_ranges(*wasm_start, *wasm_end)
                        })
```

Reply:
Courtesy of Haskell's [`concatMap`](https://hoogle.haskell.org/?hoogle=concatMap) :-)
Potentially a bit faster.